RandBLAS/trig_skops.hh

```diff
@@ -0,0 +1,315 @@
+        T* B // The destination matrix
+        // int64_t ldb
+    ) {
+        // B <- alpha * diag(Rademacher) * op(A) + beta * B
+        // `alpha`, `beta` aren't needed
+        // && shape(A) == shape(B) && lda == ldb && shape({A | B}) = m-by-n
+        //NOTE: Use `layout` and `opA` for working with RowMajor data
+        int64_t lda = m;
+        int64_t ldb = m;
+
+        //NOTE: Should the copy be made inside of `applyDiagonalRademacher` or
+        // should it be inside of `lskget`?
+        lapack::lacpy(lapack::MatrixType::General, m, n, A, lda, B, ldb);
+
+        applyDiagonalRademacher(m, n, B, diag);
+    }
+
+
+    // `permuteRowsToTop` should just take in `B` as an input
+    // `B` will be passed in by the user to `sketch_general/lskget` and `permuteRowsToTop` will take in
+    // the `B` that has been modified by `applyRademacherDiagonal`
+    // Will also be called inside of `lskget`
+    template<typename T>
+    void permuteRowsToTop(int rows, int cols, std::vector<int64_t> selectedRows, T* B, int ldb) {
+        //NOTE: There should be a similar `permuteColsToLeft`
+        int top = 0;  // Keeps track of the topmost unselected row
+
+        // Compute `r` internally
+        // int64_t r = 100;
+        // std::vector<int64_t> selectedRows = generateRademacherVector_parallel(r);
+
+        for (int selected : selectedRows) {
+            if (selected != top) {
+                // Use BLAS swap to swap the entire rows
+                // Swapping row 'selected' with row 'top'
+                blas::swap(cols, &B[selected], ldb, &B[top], ldb);
+            }
+            top++;
+        }
+    }
+
+    enum class TrigDistName: char {
+        Fourier = 'F',
+
+        // ---------------------------------------------------------------------------
+
+        Hadamard = 'H'
+    };
+
+    struct TrigDist {
+        const int64_t n_rows;
+        const int64_t n_cols;
+
+        int64_t dim_short;
+        int64_t dim_long;
+
+        const TrigDistName family;
+
+        TrigDist(
+            int64_t n_rows,
+            int64_t n_cols,
+            TrigDistName tn = TrigDistName::Hadamard
+        ) : n_rows(n_rows), n_cols(n_cols), family(tn) {
+            dim_short = MIN(n_rows, n_cols);
+            dim_long = MAX(n_rows, n_cols);
+}
+};
+
+    template<typename T, SignedInteger sint_t, typename RNG = r123::Philox4x32>
+    struct TrigSkOp {
+        using generator = RNG;
+        using state_type = RNGState<RNG>;
+        using buffer_type = T;
+
+        const int64_t n_rows;
+        const int64_t n_cols;
+        int64_t dim_short;
+        int64_t dim_long;
+        // int64_t n_sampled;
+
+        const TrigDist dist;
+
+
+        const RNGState<RNG> seed_state;
+
+        RNGState<RNG> next_state;
+
+        const blas::Layout layout;
+        const bool sketchFromLeft = true;
+        bool known_filled = false;
+
+        T* DiagonalRademacher = nullptr;
+        T* SampledRows = nullptr;
```

Comment:
1. Use snake case, not camel case.
2. sampled rows should be ``int64_t*``, not ``T*``.
3. rename "sampled rows" to something like ``subsample_idxs`` or ``sampled_coords`` or ``subsampled_coords``. This way it can refer to either rows or columns.

Reply:
Using snake_case uniformly throughout all of my code now!